src/k8s/pkg/k8sd/controllers/node_configuration.go

```diff
@@ -0,0 +1,83 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+)
+
+type NodeConfigurationController struct {
+}
+
+func NewNodeConfigurationController() *NodeConfigurationController {
+	return &NodeConfigurationController{}
+}
+
+func (c *NodeConfigurationController) Run(ctx context.Context, snap snap.Snap) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		client, err := k8s.NewNodeClient(snap)
+		if err != nil {
+			log.Println(fmt.Errorf("failed to create kubernetes node client: %w", err))
+			// This fails when the node is not bootstrapped or joined, acts like a retry timer
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
+		configCtx, cancel := context.WithCancel(ctx)
+		go client.WatchNodeConfig(configCtx, snap, c.reconcile, cancel)
+		<-configCtx.Done()
+		// This also can fail during bootstrapping/start up when api-server is not ready
+		// So the watch requests get connection refused replies
+		time.Sleep(3 * time.Second)
+	}
+}
+
+func (c *NodeConfigurationController) reconcile(ctx context.Context, snap snap.Snap, config types.NodeConfig) error {
+	var kubeletChanged bool
+
+	if config.ClusterDNS != "" {
+		var err error
+		kubeletChanged, err = snaputil.UpdateServiceArguments(snap, "kubelet", map[string]string{
+			"--cluster-dns": config.ClusterDNS,
+		}, nil)
+		if err != nil {
+			return fmt.Errorf("failed to update kubelet arguments: %w", err)
+		}
+	} else {
+		var err error
+		kubeletChanged, err = snaputil.UpdateServiceArguments(snap, "kubelet", nil, []string{"--cluster-dns"})
+		if err != nil {
+			return fmt.Errorf("failed to update kubelet arguments: %w", err)
+		}
+	}
+
+	if config.ClusterDomain != "" {
+		var err error
+		kubeletChanged, err = snaputil.UpdateServiceArguments(snap, "kubelet", map[string]string{
+			"--cluster-domain": config.ClusterDomain,
+		}, nil)
+		if err != nil {
+			return fmt.Errorf("failed to update kubelet arguments: %w", err)
+		}
+	}
+
+	if kubeletChanged {
```

Comment:
two notes here: i'd call this `mustRestartKubelet`, to be clear on the intent of how this will be used. secondly, note that this value might not be correct as the code is structured (if both clusterDNS and clusterDomain are set, only the clusterDomain is considered for restarting kubelet, which is not right.